### PR TITLE
fix output path conflict

### PR DESCRIPTION
**Pull request by AngryBerryMS**
[AB#429891](https://dev.azure.com/ceapex/Engineering/_workitems/edit/429891)

### CODE REVIEW

File: src/docfx/publish/PublishUrlMap.cs [hunk 1]
```diff
@@ -132,16 +132,34 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
             }
 
             // redirection file is preferred than source file
-            var redirection = conflicts.FirstOrDefault(x => x.SourcePath.Origin == FileOrigin.Redirection);
-            if (redirection != null)
+            var redirections = conflicts.Where(x => x.SourcePath.Origin == FileOrigin.Redirection).OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer);
+            var nonRedirections = conflicts.Where(x => x.SourcePath.Origin != FileOrigin.Redirection)
+                .OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer)
+                .Select(x => x.SourcePath.Path.Value).ToList();
```
Comment (live1206) [inline src/docfx/publish/PublishUrlMap.cs:138]: Change the input type to `IEnumerable`, `ToList()` can be removed

File: src/docfx/publish/PublishUrlMap.cs [hunk 2]
```diff
@@ -132,16 +132,34 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
...
+            var redirection = redirections.FirstOrDefault();
+            var redirectionCount = redirections?.Count() ?? 0;
```
Comment (live1206) [inline src/docfx/publish/PublishUrlMap.cs:140]: why is `?` needed here?

File: src/docfx/publish/PublishUrlMap.cs [hunk 3]
```diff
@@ -132,16 +132,34 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
...
+                return conflicts.OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer).Last();
             }
+            else if (redirectionCount == 1)
+            {
+                _errors.Add(Errors.Redirection.RedirectedFileNotRemoved(nonRedirections));
+                return redirection!;
+            }
+            else
+            {
+                if (nonRedirectionCount > 0)
```
Comment (live1206) [inline src/docfx/publish/PublishUrlMap.cs:155]: Can this be `conflicts.Count() > redirectionCount`? We can save a field declaration

File: src/docfx/publish/PublishUrlMap.cs [hunk 4]
```diff
@@ -132,16 +132,34 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
             }
 
             // redirection file is preferred than source file
-            var redirection = conflicts.FirstOrDefault(x => x.SourcePath.Origin == FileOrigin.Redirection);
-            if (redirection != null)
+            var redirections = conflicts.Where(x => x.SourcePath.Origin == FileOrigin.Redirection).OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer);
+            var nonRedirections = conflicts.Where(x => x.SourcePath.Origin != FileOrigin.Redirection)
+                .OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer)
+                .Select(x => x.SourcePath.Path.Value).ToList();
+            var redirection = redirections.FirstOrDefault();
```
Comment (live1206) [inline src/docfx/publish/PublishUrlMap.cs:139]: nit: consider put this declaration into a block where `redirectionCount > 0`, then we can get rid of all the nullable handling

Comment (AngryBerryMS) [reply src/docfx/publish/PublishUrlMap.cs:139]: There are 2 parts that satisfy `redirectionCount > 0`. I don't want too many nested `if...else...`

File: src/docfx/publish/PublishUrlMap.cs (hunk 2 above)
Comment (AngryBerryMS) [reply src/docfx/publish/PublishUrlMap.cs:140]: Fixed.

File: src/docfx/publish/PublishUrlMap.cs (hunk 1 above)
Comment (AngryBerryMS) [reply src/docfx/publish/PublishUrlMap.cs:138]: Done.

File: src/docfx/publish/PublishUrlMap.cs [hunk 5]
```diff
@@ -132,16 +132,33 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
...
+                {
+                    _errors.Add(Errors.Redirection.RedirectedFileNotRemoved(nonRedirections));
+                }
 
-            // otherwise, prefer the one based on FilePath
-            _errors.Add(Errors.UrlPath.OutputPathConflict(conflicts.First().OutputPath, conflicts.Select(x => x.SourcePath)));
-            return conflicts.OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer).Last();
+                _errors.Add(Errors.UrlPath.OutputPathConflict(conflicts.First().OutputPath, conflicts.Select(x => x.SourcePath)));
+                return redirection!;
+            }
```
Comment (mingwli) [inline src/docfx/publish/PublishUrlMap.cs:161]:
```c#
            if (redirection != null && conflicts.Any(x => x.SourcePath.Origin != FileOrigin.Redirection))
            {
                // redirection file is preferred than source file
                _errors.Add(Errors.Redirection.RedirectedFileNotRemoved(redirection.SourcePath));
                return redirection;
            }

            return conflicts.OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer).Last();
```

What if always report `output-path-conflict`, and report `redirected-file-not-removed` when conflicts contain both.
This make the logic easier to describe and maintain, and won't cause extra confusion to end user.

@live1206 @herohua @AngryBerryMS

Comment (live1206) [reply src/docfx/publish/PublishUrlMap.cs:161]: We need to inform the user that, a redirection has been added and the original file needs to be removed.
This removing action point needs to be stressed out, we had this requirement from user previously and confirmed with PM.

**Review by live1206 — APPROVED** [review]

**Pull request merged**